crates/uv/src/commands/tool/run.rs

```diff
@@ -0,0 +1,127 @@
+use std::ffi::OsString;
+use std::path::PathBuf;
+
+use anyhow::Result;
+use itertools::Itertools;
+use tempfile::tempdir_in;
+use tokio::process::Command;
+use tracing::debug;
+
+use uv_cache::Cache;
+use uv_configuration::PreviewMode;
+use uv_interpreter::PythonEnvironment;
+use uv_requirements::RequirementsSource;
+use uv_warnings::warn_user;
+
+use crate::commands::update_environment;
+use crate::commands::ExitStatus;
+use crate::printer::Printer;
+
+/// Run a command.
+#[allow(clippy::too_many_arguments)]
+pub(crate) async fn run(
+    target: String,
+    args: Vec<OsString>,
+    python: Option<String>,
+    _isolated: bool,
+    preview: PreviewMode,
+    cache: &Cache,
+    printer: Printer,
+) -> Result<ExitStatus> {
+    if preview.is_disabled() {
+        warn_user!("`uv tool run` is experimental and may change without warning.");
+    }
+
+    // TODO(zanieb): Allow users to pass an explicit package name different than the target
+    // as well as additional requirements
+    let requirements = [RequirementsSource::from_package(target.clone())];
+
+    // TODO(zanieb): When implementing project-level tools, discover the project and check if it has the tool
+    // TOOD(zanieb): Determine if we sould layer on top of the project environment if it is present
+
+    // If necessary, create an environment for the ephemeral requirements.
+    debug!("Syncing ephemeral environment.");
+
+    // Discover an interpreter.
+    let interpreter = if let Some(python) = python.as_ref() {
+        PythonEnvironment::from_requested_python(python, cache)?.into_interpreter()
+    } else {
+        PythonEnvironment::from_default_python(cache)?.into_interpreter()
+    };
+
+    // Create a virtual environment
+    // TODO(zanieb): Move this path derivation elsewhere
+    let uv_state_path = std::env::current_dir()?.join(".uv");
+    fs_err::create_dir_all(&uv_state_path)?;
+    let tmpdir = tempdir_in(uv_state_path)?;
+    let venv = uv_virtualenv::create_venv(
+        tmpdir.path(),
+        interpreter,
+        uv_virtualenv::Prompt::None,
+        false,
+        false,
+    )?;
```

Comment:
Ultimately, will these be installed in some sort of globally directory that's also exposed on PATH, either by adding at install time or via some kind of `ensurepath`?

Reply:
Ok, I see this in https://github.com/astral-sh/uv/issues/3560 now. If the tool is already installed in the user environment, it will be used, but `uv tool run` on its own will not _install_ the tool. Is that correct?

Reply:
Yeah `uv tool run` will not install a tool. It could cache the environment and will support detecting user-installed and project-level tools but we don't have those yet here. I've already started a branch for `uv tool install` — it's just much more complex as a starting point.